static/app/views/dashboards/widgetCard/autoSizedText.tsx

```diff
@@ -0,0 +1,177 @@
+import {useLayoutEffect, useRef, useState, useTransition} from 'react';
+import styled from '@emotion/styled';
+import {useResizeObserver} from '@react-aria/utils';
+import debounce from 'lodash/debounce';
+
+interface Props {
+  children: React.ReactNode;
+  maxFontSize: number;
+  minFontSize: number;
+  calculationCountLimit?: number;
+}
+
+export function AutoSizedText({
+  children,
+  minFontSize,
+  maxFontSize,
+  calculationCountLimit = DEFAULT_CALCULATION_COUNT_LIMIT,
+}: Props) {
+  const [_isPending, startTransition] = useTransition();
+
+  const [parentHeight, setParentHeight] = useState<number | null>(null);
+  const [parentWidth, setParentWidth] = useState<number | null>(null);
+
+  const [fontSize, setFontSize] = useState<number>((maxFontSize + minFontSize) / 2);
+  const [fontSizeLowerBound, setFontSizeLowerBound] = useState<number>(minFontSize);
+  const [fontSizeUpperBound, setFontSizeUpperBound] = useState<number>(maxFontSize);
+
+  const [calculationCount, setCalculationCount] = useState<number>(0);
```

Comment:
Have you considered using a reducer for this logic? I see on L36t that there seems to be some dependency between these values, a reducer might help you express them more clearly